keps/sig-architecture/3784-database-of-names/README.md

```diff
@@ -0,0 +1,445 @@
+
+const (
+    SIGNetwork = SIG("Network")
+    SIGNode = SIG("Node")
+    SIGApps = SIG("Apps")
+    SIGStorage = SIG("Storage")
+    SIGArchitecture = SIG("Architecture")
+    // more...
+)
+
+// Context describes where a name might be used.
+type Context string
+
+const (
+    ContextLabel = Context("Label")             // key+value
+    ContextAnnotation = Context("Annotation")   // key+value
+    ContextTaint = Context("Taint")             // key
+    ContextCondition = Context("Condition")     // key
+    ContextAppProtocol = Context("AppProtocol") // key
+    ContextAppProtocol = Context("ApiGroup")    // key
+    // more...
+)
+
+// ValueType describes the datatype of a value for a key name.
+type ValueType string
+
+const (
+    ValueTypeInt = ValueType("Int")
+    ValueTypeString = ValueType("String")
+    ValueTypeBool = ValueType("Bool")
+    ValueTypeJSONObject = ValueType("JSONObject")
+    ValueTypeJSONList = ValueType("JSONList")
+    // more...
+)
+```
+
+We can publish this schema as Go code or something else if there is a better or
+more standard notation.  We can validate all names against this schema.
+
+<<[UNRESOLVED enumerate resources ]>>
+Should we enumerate resources?  Good for validation, but a bit tedious.
+<<[/UNRESOLVED]>>
+
+#### Why YAML
+
+YAML is the lingua franca of Kubernetes.  We propose to use a subset of YAML
+which is closer to JSON and easier to avoid whitespace errors, while still
+allowing comments.
+
+#### Examples
+
+/by-name/io/kubernetes/names.yaml:
+```yaml
+[{
+  key: "kubernetes.io/hostname",
+  sig: "Node",
+  context: "Label",
+  valueType: "String",
+  resources: [ "core/nodes" ],
+  description: "The name of the node.",
+}, {
+  key: "kubernetes.io/arch",
+  sig: "Node",
+  context: "Label",
+  valueType: "String",
+  resources: [ "core/nodes" ],
+  description: "The architecture of the node.",
+}, {
+  key: "kubernetes.io/os",
+  sig: "Node",
+  context: "Label",
+  valueType: "String",
+  resources: [ "core/nodes" ],
+  description: "The OS of the node.",
+}]
+```
+
+
+/by-name/io/kubernetes/topology/names.yaml:
+```yaml
+[{
+  key: "kubernetes.io/hostname",
+  key: "topology.kubernetes.io/zone",
+  sig: "Architecture",
+  context: "Label",
+  valueType: "String",
+  resources: [ "core/nodes", core/persistentvolumes ],
+  description: "The availability zone of the resource.",
+  docsLink: https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesiozone
+}, {
+  key: "topology.kubernetes.io/region",
+  sig: "Architecture",
+  context: "Label",
+  valueType: "String",
+  resources: [ "core/nodes", core/persistentvolumes ],
+  description: "The availability region of the resource.",
+  docsLink: https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesioregion
+}]
+```
+
+<<[UNRESOLVED key collisions ]>>
+Should we have a YAML per key - easier lookups and collision avoidance, but
+lots of files.  Would maybe need Context to be a list.
```

Comment:
Another potential home for one-file-per-key is k/website. Code that wants to could pull from the website repo and turn [front matter](https://gohugo.io/content-management/front-matter/) YAML into something that other tools want to consume.

Reply:
I'd like ONE source of truth and the rest derived from that, if possible.  Secondary concern is ease of adding/updating (which is why I suggest k/k/staging).  But I am open...

Reply:
Website is the current source of truth - but not machine readable.

Reply:
If we put enough human-friendly text into this DB, would you want to generate the doc from it?